Sources/OpenTelemetry/Processing/BatchSpanProcessor.swift

```diff
@@ -0,0 +1,80 @@
+//
+// See LICENSE.txt for license information
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOConcurrencyHelpers
+
+extension OTel {
+    /// A processor which hands off ended spans to a given exporter in batches.
+    ///
+    /// - SeeAlso: [OTel Spec: Batching processor](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#batching-processor)
+    public final class BatchSpanProcessor: OTelSpanProcessor {
+        private let exporter: OTelSpanExporter
+        private let eventLoopGroup: EventLoopGroup
+        private let maxQueueSize: Int
+        private let maxBatchSize: Int
+
+        private var queue = CircularBuffer<OTel.RecordedSpan>()
+        private let queueLock = Lock()
+        private var exportTask: RepeatedTask!
+
+        /// Initialize a batch span processor.
+        /// - Parameters:
+        ///   - maxBatchSize: The maximum number of spans exported in one batch. Defaults to `512`.
+        ///   - maxQueueSize: The maximum number of spans queued for export. Defaults to `2048`.
+        ///   Old spans will be dropped if the number of queued spans exceeds this max size.
+        ///   - interval: The time interval in which to export batches. Defaults to 5 seconds.
+        ///   - exporter: The exporter which receives processed span batches.
+        ///   - eventLoopGroup: The event loop group on which to process.
+        public init(
+            maxBatchSize: Int = 512,
+            maxQueueSize: Int = 2048,
+            interval: TimeAmount = .seconds(5),
+            exportingTo exporter: OTelSpanExporter,
+            eventLoopGroup: EventLoopGroup
+        ) {
+            self.exporter = exporter
+            self.eventLoopGroup = eventLoopGroup
+            self.maxBatchSize = maxBatchSize
+            self.maxQueueSize = maxQueueSize
+            exportTask = eventLoopGroup.next().scheduleRepeatedAsyncTask(
+                initialDelay: interval,
+                delay: interval,
+                exportBatch
+            )
+        }
+
+        public func processEndedSpan(_ span: OTel.RecordedSpan) {
+            queueLock.withLock {
+                if queue.count == maxQueueSize {
+                    queue.removeFirst()
+                }
+                queue.append(span)
+            }
+        }
+
+        public func shutdownGracefully() -> EventLoopFuture<Void> {
+            let promise = eventLoopGroup.next().makePromise(of: Void.self)
+            exportTask.cancel(promise: promise)
+            return promise.futureResult
+        }
+
+        private func exportBatch(_ task: RepeatedTask) -> EventLoopFuture<Void> {
+            queueLock.withLock {
+                let spans = queue.prefix(maxBatchSize)
+                queue.removeFirst(spans.count)
+                return exporter.export(spans)
+            }
+        }
+    }
+}
```

Comment:
okey that's a good start :) we can try more fine grained locking in the future but this is a good start (only take a lock to swap `[]` with the current ones, and then flush the obtained one)

Reply:
(or by putting a real concurrent queue in here :-))

Reply:
Sounds good - I'll keep that in mind when revisiting at some point 👍